### DISCUSSION TITLE
[question] Can we make TUR be similar to NUR?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [termux-user-repository/tur#872] [Ideas]
<https://github.com/nix-community/NUR> collect third-party packages of NixOS, which is a combination of many users' github repositories named nur-packages. Can we make TUR be similar to NUR? Just combine many users' github repositories.

**Comment by T-Dynamos** [comment]
> many users' github repositories

Which repos you mean by? Examples?

**Comment by licy183** [comment]
Sadly, maybe not. IIUC NUR (and its repos) contains lots of scripts for users and users can compile these packages on their device. Yeah, it is possible for TUR to provide such a function. But, Termux is often used on arm devices, and compiling will cost lots of time. Actually, TUR wants to provide prebuilt binaries for users to install and save building time on arm devices.

**Comment by Freed-Wu** [comment]
> Which repos you mean by? Examples?

For example, <https://nur.nix-community.org> combines 199 user repositories currently.

> compiling will cost lots of time

Oh, I know many nur-packages use [cachix](https://app.cachix.org/) to provide binary cache. I don't know any methods to provide binary cache for termux packages. :cry:

**Comment by licy183** [comment]
Termux uses apt (or pacman), and it is much different from NixOS.

**Comment by Freed-Wu** [comment]
NUR is a combination of many users' nur-packages. There exists a CI/CD in NUR repo to combine all users' nur-package repos to one repo. every user's packages is under a namespace named `nur.user_name` to avoid conflict.

AUR is another method. AUR provide a git platform, every one can create package in AUR. once one user create repo `foo` for software foo, other user can contact him/her to send patch. create different package for same software is disencouraged.

Currently, TUR have a small amount of packages compared to AUR, so use a git repo to collect all user's packages is enough.

**Comment by Freed-Wu** [comment]
perhaps TUR can be organized like Gentoo overlay (Or like Ubuntu PPA), that is every packager have its overlay repo, user select to install whose overlay repos. When two packagers' overlay repos have two packages with same name, user select install whose package. It is differen from AUR and NUR, which also solve the conflict problem without NUR's namespace or AUR's unified repo. Just my 2c.

Maybe transfer this issue to discussion will be better? Sorry for finding discussion so late :cry:

**Comment by Freed-Wu** [comment]
So is it possible to use it as a project template to create a customized source, user just add it to `$PREFIX/etc/apt/sources.list.d/XXX.list` to use it?

**Reply by licy183** [reply]
Of course it's possible, but it might take a lot of work.

TUR hosts all the binary debs in GitHub Release, and use a script (https://github.com/termux-user-repository/dists/blob/master/scripts/cf-worker.js) to redirect pages like an APT repository. If someone wants to setup an APT repository like TUR does, they should fork the [dists repository](https://github.com/termux-user-repository/dists) and create Cloudflare Worker.

**Reply by Freed-Wu** [reply]
Yes, just like Ubuntu PPA, every one has one repo for theirself. They even can host **private** projects on their termux's "PPA". Although there are few document about it, It should be possible in theory.

**Reply by Freed-Wu** [reply]
And apt source seems to have binary source `deb` and source code source `deb-src`. Is it possible to let user to host `deb-src` to build the package from source? Although document is fewer, I guess it is possible?

```
deb http://security.ubuntu.com/ubuntu/ jammy-security main restricted universe multiverse
deb-src http://security.ubuntu.com/ubuntu/ jammy-security main restricted universe multiverse
```

**Reply by licy183** [reply]
Hosting `deb-src` makes no sense to Termux. Packages on Termux doesn't use `debian/control` file when building a package.

**Reply by Freed-Wu** [reply]
> Hosting deb-src makes no sense to Termux

How regretful. Such as termux doesn't allow any package is larger than 100 MB.

- https://github.com/termux/termux-packages/issues/11834
- https://github.com/termux/termux-packages/issues/10191

If termux allow build package from source, user can build it by themselves without the limitation of package size.

**Reply by licy183** [reply]
> If termux allow build package from source, user can build it by themselves without the limitation of package size.

It is possible to write an installer for these kind of packages. Ubuntu has `ttf-mscorefonts-installer` to install msfonts in `postinst` script.
